File: python/plan/test_path_smoother.py
import numpy as np
from path_smoother import PathSmoother
from map3d import Map3D
from astar3d import AStar3D


def test_basic_smoothing():
    """
    测试基本路径平滑功能
    """
    print("测试基本路径平滑功能")
    
    # 创建一个锯齿状路径
    path = [
        (0.0, 0.0, 0.0),
        (1.0, 0.0, 0.0),
        (1.0, 1.0, 0.0),
        (2.0, 1.0, 0.0),
        (2.0, 2.0, 0.0),
        (3.0, 2.0, 0.0),
        (3.0, 3.0, 0.0),
        (4.0, 3.0, 0.0),
        (4.0, 4.0, 0.0)
    ]
    
    print(f"原始路径点数: {len(path)}")
    
    # 创建路径平滑器
    smoother = PathSmoother()
    
    # 平滑路径
    smoothed_path = smoother.smooth_path(path, smoothing_factor=0.5, num_points=50)
    
    print(f"平滑后路径点数: {len(smoothed_path)}")
    print(f"前5个平滑路径点: {smoothed_path[:5]}")
    print(f"后5个平滑路径点: {smoothed_path[-5:]}")
    
    return path, smoothed_path


def test_astar_path_smoothing():
    """
    测试A*路径的平滑处理
    """
    print("\n\n测试A*路径的平滑处理")
    
    # 创建地图
    map3d = Map3D(20, 20, 20, 0.5)
    
    # 添加一些障碍物
    for i in range(5, 15):
        map3d.add_obstacle(2.5, i * 0.5, 2.5)
        map3d.add_obstacle(5.0, i * 0.5, 2.5)
    
    # 创建A*算法实例
    astar = AStar3D(map3d)
    
    # 规划路径
    start = (0.0, 0.0, 0.0)
    goal = (9.5, 9.5, 9.5)
    
    print(f"路径规划: 从 {start} 到 {goal}")
    path = astar.plan(start, goal)
    
    if not path:
        print("未找到路径")
        return None, None
    
    print(f"原始路径点数: {len(path)}")
    
    # 平滑路径（带障碍物避让）
    smoother = PathSmoother()
    smoothed_path = smoother.smooth_path_with_obstacle_avoidance(path, map3d, smoothing_factor=1.0)
    
    print(f"平滑后路径点数: {len(smoothed_path)}")
    
    # 检查平滑路径是否与障碍物冲突
    conflict_count = 0
    for point in smoothed_path:
        if map3d.is_obstacle(point[0], point[1], point[2]):
            conflict_count += 1
    
    print(f"平滑路径与障碍物冲突点数: {conflict_count}")
    
    return path, smoothed_path


def compare_smoothing_factors():
    """
    比较不同平滑因子的效果
    """
    print("\n\n比较不同平滑因子的效果")
    
    # 创建一个弯曲路径
    path = []
    for i in range(20):
        t = i / 19.0 * 2 * np.pi
        x = t
        y = np.sin(t)
        z = np.cos(t)
        path.append((x, y, z))
    
    smoother = PathSmoother()
    
    # 测试不同的平滑因子
    factors = [0.1, 0.5, 1.0, 2.0]
    for factor in factors:
        smoothed_path = smoother.smooth_path(path, smoothing_factor=factor, num_points=100)
        print(f"平滑因子 {factor}: 生成 {len(smoothed_path)} 个点")


def test_obstacle_avoidance():
    """
    测试障碍物避让功能
    """
    print("\n\n测试障碍物避让功能")
    
    # 创建地图
    map3d = Map3D(20, 20, 20, 0.5)
    
    # 添加一些障碍物形成一个通道
    for i in range(8, 12):
        for j in range(8, 12):
            map3d.add_obstacle(i * 0.5, j * 0.5, 2.5)
    
    # 创建一个穿过障碍物区域的路径
    path = [
        (0.0, 0.0, 0.0),
        (1.0, 1.0, 1.0),
        (2.0, 2.0, 2.0),
        (3.0, 3.0, 3.0),
        (4.0, 4.0, 4.0)
    ]
    
    print(f"原始路径点数: {len(path)}")
    
    # 检查原始路径是否与障碍物冲突
    original_conflicts = 0
    for point in path:
        if map3d.is_obstacle(point[0], point[1], point[2]):
            original_conflicts += 1
    
    print(f"原始路径与障碍物冲突点数: {original_conflicts}")
    
    # 平滑路径（带障碍物避让）
    smoother = PathSmoother()
    smoothed_path = smoother.smooth_path_with_obstacle_avoidance(path, map3d, smoothing_factor=0.5, num_points=30)
    
    print(f"平滑后路径点数: {len(smoothed_path)}")
    
    # 检查平滑路径是否与障碍物冲突
    smoothed_conflicts = 0
    for point in smoothed_path:
        if map3d.is_obstacle(point[0], point[1], point[2]):
            smoothed_conflicts += 1
    
    print(f"平滑路径与障碍物冲突点数: {smoothed_conflicts}")
    
    return path, smoothed_path


def test_improved_features():
    """
    测试改进的功能：启发式搜索和局部再平滑
    """
    print("\n\n测试改进的功能：启发式搜索和局部再平滑")
    
    # 创建地图
    map3d = Map3D(20, 20, 20, 0.5)
    
    # 添加一个障碍物块
    for x in range(8, 12):
        for y in range(8, 12):
            for z in range(8, 12):
                map3d.add_obstacle(x * 0.5, y * 0.5, z * 0.5)
    
    # 创建一条穿过障碍物中心的路径
    path = [
        (1.0, 1.0, 1.0),
        (2.0, 2.0, 2.0),
        (3.0, 3.0, 3.0),
        (4.0, 4.0, 4.0),
        (5.0, 5.0, 5.0)
    ]
    
    print(f"原始路径点数: {len(path)}")
    
    # 检查原始路径是否与障碍物冲突
    original_conflicts = 0
    for point in path:
        if map3d.is_obstacle(point[0], point[1], point[2]):
            original_conflicts += 1
    
    print(f"原始路径与障碍物冲突点数: {original_conflicts}")
    
    # 使用改进的平滑算法
    smoother = PathSmoother()
    smoothed_path = smoother.smooth_path_with_obstacle_avoidance(
        path, map3d, smoothing_factor=0.8, num_points=40
    )
    
    print(f"改进算法平滑后路径点数: {len(smoothed_path)}")
    
    # 检查平滑路径是否与障碍物冲突
    smoothed_conflicts = 0
    for point in smoothed_path:
        if map3d.is_obstacle(point[0], point[1], point[2]):
            smoothed_conflicts += 1
    
    print(f"改进算法平滑路径与障碍物冲突点数: {smoothed_conflicts}")
    
    return path, smoothed_path


if __name__ == "__main__":
    # 运行测试
    test_basic_smoothing()
    test_astar_path_smoothing()
    compare_smoothing_factors()
    test_obstacle_avoidance()
    test_improved_features()
    
    print("\n所有测试完成")

File: cpp/test_path_smoother.cpp
#include "PathSmoother.h"
#include "AStar3D.h"
#include <iostream>
#include <memory>

int main() {
    try {
        // 创建一个较小的3D地图实例
        auto map3d = std::make_shared<Map3D>(10, 10, 10, 1.0, 0.0, 0.0, 0.0);
        
        // 设置膨胀半径为0，避免过度膨胀
        map3d->setInflationRadius(0.0);
        
        // 添加一些障碍物
        map3d->addObstacle(2.0, 2.0, 2.0);
        map3d->addObstacle(2.0, 2.0, 3.0);
        map3d->addObstacle(2.0, 3.0, 2.0);
        
        // 创建A*算法实例
        AStar3D astar(map3d);
        
        // 设置起始点和目标点
        std::tuple<double, double, double> start(0.0, 0.0, 0.0);
        std::tuple<double, double, double> goal(5.0, 5.0, 5.0);
        
        std::cout << "开始A*路径规划..." << std::endl;
        std::cout << "起点: (" << std::get<0>(start) << ", " << std::get<1>(start) << ", " << std::get<2>(start) << ")" << std::endl;
        std::cout << "终点: (" << std::get<0>(goal) << ", " << std::get<1>(goal) << ", " << std::get<2>(goal) << ")" << std::endl;
        
        // 执行路径规划
        auto path = astar.plan(start, goal);
        
        if (!path.empty()) {
            std::cout << "找到路径，路径长度: " << path.size() << " 个节点" << std::endl;
            
            // 创建路径平滑器
            PathSmoother smoother;
            
            std::cout << "开始路径平滑..." << std::endl;
            
            // 使用路径平滑器
            auto smoothed_path = smoother.smooth_path(path, 0.5, 20);
            
            std::cout << "平滑后路径长度: " << smoothed_path.size() << " 个节点" << std::endl;
            
            // 显示原始路径的前几个点和平滑后路径的前几个点
            std::cout << "\n原始路径前5个点:" << std::endl;
            for (size_t i = 0; i < std::min(static_cast<size_t>(5), path.size()); ++i) {
                auto [x, y, z] = path[i];
                std::cout << "  [" << i << "] (" << x << ", " << y << ", " << z << ")" << std::endl;
            }
            
            std::cout << "\n平滑后路径前5个点:" << std::endl;
            for (size_t i = 0; i < std::min(static_cast<size_t>(5), smoothed_path.size()); ++i) {
                auto [x, y, z] = smoothed_path[i];
                std::cout << "  [" << i << "] (" << x << ", " << y << ", " << z << ")" << std::endl;
            }
            
            std::cout << "\n平滑后路径后5个点:" << std::endl;
            for (size_t i = std::max(static_cast<size_t>(0), smoothed_path.size() - 5); i < smoothed_path.size(); ++i) {
                auto [x, y, z] = smoothed_path[i];
                std::cout << "  [" << i << "] (" << x << ", " << y << ", " << z << ")" << std::endl;
            }
            
            std::cout << "\n开始带障碍物避让的路径平滑..." << std::endl;
            
            // 使用带障碍物避让的路径平滑
            auto obstacle_avoiding_path = smoother.smooth_path_with_obstacle_avoidance(path, map3d, 0.5, 20, 1.0, 5);
            
            std::cout << "避障平滑后路径长度: " << obstacle_avoiding_path.size() << " 个节点" << std::endl;
            
            std::cout << "\n避障平滑后路径前5个点:" << std::endl;
            for (size_t i = 0; i < std::min(static_cast<size_t>(5), obstacle_avoiding_path.size()); ++i) {
                auto [x, y, z] = obstacle_avoiding_path[i];
                std::cout << "  [" << i << "] (" << x << ", " << y << ", " << z << ")" << std::endl;
            }
            
            std::cout << "\n避障平滑后路径后5个点:" << std::endl;
            for (size_t i = std::max(static_cast<size_t>(0), obstacle_avoiding_path.size() - 5); i < obstacle_avoiding_path.size(); ++i) {
                auto [x, y, z] = obstacle_avoiding_path[i];
                std::cout << "  [" << i << "] (" << x << ", " << y << ", " << z << ")" << std::endl;
            }
        } else {
            std::cout << "未找到路径!" << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cerr << "发生错误: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}